src/rgw/rgw_data_sync.cc

```diff
@@ -1698,7 +1704,7 @@ class RGWDataSyncShardCR : public RGWCoroutine {
             }
             tn->log(20, SSTR("handle error entry key=" << error_marker << " timestamp=" << entry_timestamp));
             spawn(sync_single_entry(source_bs, error_marker, "",
-                                    entry_timestamp, true), false);
+                                    entry_timestamp, true, nullopt), false);
```

Comment:
@smanjara I'm particularly unsure about this spot. Obviously we shouldn't include the generation number in the error repo entry since it'll just fail all over again, but I'm not sure if we should unconditionally retry it either.

Reply:
It looks right to me. @cbodley thoughts?

Reply:
the error_repo does need to track generation numbers

after a reshard, we'll start seeing datalog entries with the new generation number. but we can't process them until we've finished with the previous generation, so we need to store the new generation's entry in the error_repo for retry

sync also needs the ability to retry sync of bucket shards in the previous generation that it's still working on

Reply:
@cbodley in the datalog entries, don't we need a log generation number rather than the index generation number?

Reply:
> @cbodley in the datalog entries, don't we need a log generation number rather than the index generation number?

yes, good catch! this pr is using index layout generation numbers (`bucket_info.layout.current_index.gen`), where it should be using a log layout generation

can we change the `add_entry()` interface to take `const bucket_log_layout_generation& gen` instead of `uint64_t gen_id` to avoid this confusion? then callers would just pass `bucket_info.layout.logs.back()`